src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,296 @@
+                MethodDeclarationSyntax methodDecl => methodDecl.Body,
+                ConstructorDeclarationSyntax constructorDecl => constructorDecl.Body,
+                LocalFunctionStatementSyntax localFunctionStatement => localFunctionStatement.Body,
+                AnonymousFunctionExpressionSyntax anonymousFunction => anonymousFunction.Block,
+                OperatorDeclarationSyntax operatorDecl => operatorDecl.Body,
+                ConversionOperatorDeclarationSyntax conversionDecl => conversionDecl.Body,
+                _ => throw ExceptionUtilities.UnexpectedValue(node)
+            };
+
+            // More scenarios should be supported eventually: https://github.com/dotnet/roslyn/issues/58699
+            if (block is null)
+            {
+                return;
+            }
+
+            var methodSymbol = node is AnonymousFunctionExpressionSyntax
+                ? (IMethodSymbol?)semanticModel.GetSymbolInfo(node, cancellationToken).Symbol
+                : (IMethodSymbol?)semanticModel.GetDeclaredSymbol(node, cancellationToken);
+            if (methodSymbol is null || methodSymbol.Parameters.IsEmpty)
+            {
+                return;
+            }
+
+            foreach (var statement in block.Statements)
+            {
+                var parameter = TryGetParameterNullCheckedByStatement(statement);
+                if (parameter is not null
+                    && (!parameter.Type.IsValueType
+                        || parameter.Type.OriginalDefinition.SpecialType == SpecialType.System_Nullable_T
+                        || parameter.Type.TypeKind is TypeKind.Pointer or TypeKind.FunctionPointer)
+                    && parameter.RefKind == RefKind.None
+                    && parameter.DeclaringSyntaxReferences.FirstOrDefault() is SyntaxReference reference
+                    && reference.SyntaxTree.Equals(statement.SyntaxTree)
+                    && reference.GetSyntax() is ParameterSyntax parameterSyntax)
+                {
+                    context.ReportDiagnostic(DiagnosticHelper.Create(
+                        Descriptor,
+                        statement.GetLocation(),
+                        option.Notification.Severity,
+                        additionalLocations: new[] { parameterSyntax.GetLocation() },
+                        properties: null));
+                }
+            }
+
+            return;
+
+            IParameterSymbol? TryGetParameterNullCheckedByStatement(StatementSyntax statement)
+            {
+                switch (statement)
+                {
+                    // if (param == null) { throw new ArgumentNullException(nameof(param)); }
+                    // if (param is null) { throw new ArgumentNullException(nameof(param)); }
+                    // if (object.ReferenceEquals(param, null)) { throw new ArgumentNullException(nameof(param)); }
+                    case IfStatementSyntax ifStatement:
+                        ExpressionSyntax left, right;
```

Comment:
i would extract the entire if-handling code to it's own helper.

Reply:
I think I won't make this change due to lack of uniformity with the ExpressionStatement case. See https://github.com/dotnet/roslyn/pull/58182#discussion_r782508142